### DISCUSSION TITLE
Setting a single bit in USART control register

### DISCUSSION THREAD

**Discussion by elenziat** [Rahix/avr-hal#486] [Q&A] [answered]
Maybe I have just a wrong understanding of how the API should work.

I expected that the follwowing two statements result in the same state of the USCR0B register: (`self.usart` is of type `avr_device::atmega328p::USART0`
```rust
self.usart.ucsr0b.write(|w| w.udrie0().set_bit());
```
```rust
self.usart
    .ucsr0b
    .modify(|r, w| unsafe { w.bits(r.bits() | (1 << 5)) });
```
However, the first variant seems to clear all other bits in the register. Is this expected, or is it a weird problem within my setup?

**Comment by Rahix** [comment, marked as answer]
`.write()` and `.modify()` do very different things.  `.write()` writes the whole register where all bits start at their reset value before being modified by the closure.  See <https://docs.rs/svd2rust/latest/svd2rust/#write>. `.modify()` is a read-modify-write cycle where the current register value is read, then modified by the closure, and then the modified value is written back to the register.  Check <https://docs.rs/svd2rust/latest/svd2rust/#modify>.

Now there is one additional detail you should know for AVR in particular.  For some registers, a special instruction can be used to just set or clear a single bit.  The compiler is smart enough to optimize a `.modify()` call which only sets a single bit to use this instruction.

**Reply by elenziat** [reply]
Thanks!

I have misinterpreted the "or write only the fields you need" in the doc `.write()` given in this crate: [https://rahix.github.io/avr-hal/avr_device/generic/struct.Reg.html#method.write]()
~~Especially the difference of the last two of the three examples is unclear to me. (The last example explicitly states that the other fields will be reset to the default value. From what you have written in you answer, this is also the case for the example above, isn't?~~ Ok, there are only two examples, the second having two variants, now I get it.

If I understand it now correctly, the difference to `.write_with_zeros()` [https://rahix.github.io/avr-hal/avr_device/generic/struct.Reg.html#method.write_with_zero]() is that `.write_with_zeros()` sets the unused fields to 0 and `.write()` uses the default values?

**Reply by Rahix** [reply]
Exactly.  `.write()` will only differ from `.write_with_zeros()` if there are some fields with a non-zero reset value in the register.  This isn't terribly common though, most fields have 0 as their reset value.
